ruby/private/binary_wrapper.tpl

```diff
@@ -1,3 +1,8 @@
+#shell #!/usr/bin/env bash
+#shell # This conditional is evaluated as true in Ruby, false in shell
+#shell if [ ]; then
```

Comment:
This is a very interesting trick! 

Perhaps I may not fully understand why this is needed, but may I offer an alternative solution to a bilingual script that might have some advantages (at the very least nothing needs to be stripped out from it):

Pure Ruby file until `__END__`, and then pure shell script after.

#### File: `wrapper.tpl`

This file contains Ruby and Shell combined, separated by `__END__` as you suggested:

```ruby
#!/usr/bin/env ruby
# vi: ft=ruby
@h = { name: 'Konstantin', company: 'Coinbase' }
require 'json'
puts JSON.pretty_generate(@h)

__END__
#!/usr/bin/env bash
# vi: ft=sh

function num-procs() {
  echo -n "Counting live processes on this $(uname -s): "
  ps -ef | wc -l | tr -d ' '
}
num-procs
```

One way to executing an appropriate one is this tiny `awk` script:

#### File `splitter.awk`

```awk
BEGIN { ruby = 1 }
{ 
  if ($1 == "__END__") {
    ruby = 0
    if (type == "ruby") { exit 0 }
    next
  }
  if (ruby == 1 && type == "ruby")  print $0
  if (ruby == 0 && type == "shell") print $0
}
```

### Extracting Ruby or Shell is easy:

We just set the AWK variable `type` to `ruby`:

```bash
❯ awk -v type=ruby -f splitter.awk bilingual.tpl 
#!/usr/bin/env ruby
# vi: ft=ruby
@h = { name: 'Konstantin', company: 'Coinbase' }
require 'json'
puts JSON.pretty_generate(@h)
```

OR `shell`

```bash
❯ awk -v type=shell -f splitter.awk bilingual.tpl 
#!/usr/bin/env bash
# vi: ft=sh
function num-procs() {
  echo -n "Counting live processes on this $(uname -s): "
  ps -ef | wc -l | tr -d ' '
}
num-procs
```

### Execution

```bash
❯ awk -v type=shell -f splitter.awk bilingual.tpl > runner; chmod 755 runner && ./runner
Counting live processes on this Darwin: 548

❯ awk -v type=ruby -f splitter.awk bilingual.tpl > runner; chmod 755 runner && ./runner
{
  "name": "Konstantin",
  "company": "Coinbase"
}
```

What I like about this approach is that we are not polluting the source template with extraneous comments that are making it hard to parse by developers.  Please let me know if this is a workable solution, or if I am missing something.

Reply:
1. It wouldn't work in `bazel run`. The target in `bazel run` must generate an executable which can be the first argument in execve(2).
  * We could add your extraction script so that the script could be the first argument on exec.  However, it doesn't actually help because the script itself must be in this same file to keep the number of the output file equal to 1.
2. It could work when you directly run the generated binary without `bazel run`. But it means you have to type `awk -v type=ruby -f splitter.awk bazel-bin/ruby/apps/hello-world-web > runner; chmod 755 runner && ./runner` instead of just `bazel-binruby/apps/hello-world-web`. I don't think it is acceptable for users.

Actually I don't think the current approach is too bad to maintain because:
* This is a common technique seen in the parser implementation in Ruby. I know it works for long term.
* Except for the `#shell ` comment and some `{var}` placeholders to be expanded by Bazel, this is a valid Ruby script.
  * The `#shell ` part is borrowed from a stable API for `runfiles.bash`. So we rarely need to change the part.

Reply:
To clarify, I'll list some requirements which this file must satisfy.

1. This file is a template file which `binary.bzl` uses when it generates an executable file corresponding to a `rb_binary` rule
2. It is better to keep the number of output files `rb_binary` equal to 1 for ease of combination with `sh_binary` rules.
3. The file generated from this template must be an executable which execve(2) accepts.
4. The initial interpreter for the generated file must not be ruby because we cannot expect that there's a ruby installed on the runtime environment when we bundle Sorbets' binary in a container image. Bazel rules usually use bash as an interpreter in this kind of cases.
5. Technically it was possible to simply write entire this script in bash script, but I wanted to manipulate Ruby's load paths in Ruby for the reasons I listed in the PR description.

As a result, the requirements lead the following restrictions:
* The file generated from this script must be valid both as a shell script and as a Ruby script
* Effectively we cannot rely on awk at *runtime*. What we can do at runtime is just execute the generated file with a user-specified list of arguments.
* It is totally fine to rely on *grep*, *sed* or *awk* at *bulid time*, though.

Reply:
Got it! Thanks for the explanation!